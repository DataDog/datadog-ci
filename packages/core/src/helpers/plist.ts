import {readFileSync} from 'fs'

import {XMLParser, XMLValidator} from 'fast-xml-parser'

type PlistNode = {[tagName: string]: [{[valueType: string]: string | number | PlistNode}]}

const parseIfEnvVariable = (propertyValue: string | number, metadata: {propertyName: string}): string | number => {
  if (typeof propertyValue !== 'string') {
    return propertyValue
  }

  // Matches "$(ANY_VARIABLE_NAME)", capturing "ANY_VARIABLE_NAME"
  const matchedEnvVariable = propertyValue.match(/^\$\((.*)\)$/)
  if (!matchedEnvVariable) {
    return propertyValue
  }

  // matchedEnvVariable[0] is the matched string, i.e. "$(VARIABLE_NAME)"
  // matchedEnvVariable[1] is the captured group, i.e. "VARIABLE_NAME"
  const value = process.env[matchedEnvVariable[1]]
  // If we haven't captured, the value is not an env variable and we should return it directly
  if (value !== undefined) {
    return value
  }

  throw new Error(`Environment variable ${matchedEnvVariable[0]} for key ${metadata.propertyName} wasn't found.`)
}

class PlistContent {
  private content: PlistNode[]
  constructor(content: PlistNode[]) {
    this.content = content
  }

  public getContent = () => this.content

  /**
   *
   * @param propertyName
   * @returns
   */
  public getPropertyValue = (propertyName: string): string | number => {
    const propertyNodeIndex = this.content.findIndex((node) => {
      if (!this.isKeyNode(node)) {
        return false
      }

      return node.key[0]['#text'] === propertyName
    })

    if (propertyNodeIndex === -1) {
      throw new Error('Property not found')
    }

    const valueNode = this.content[propertyNodeIndex + 1]
    if (!this.isStringNode(valueNode)) {
      throw new Error('Property is not a string, this is not supported yet')
    }

    return parseIfEnvVariable(valueNode.string[0]['#text'], {propertyName})
  }

  /**
   * A key entry will be turned into element of the content looking like this:
   * ```
   * {
   *   "key": [
   *     {
   *       "#text": "CFBundleDevelopmentRegion",
   *     },
   *   ],
   * }
   * ```
   */
  private isKeyNode = (node: PlistNode): node is {key: [{'#text': string}]} => {
    return !!node.key
  }

  /**
   * A string entry will be turned into element of the content looking like this:
   * ```
   * {
   *   "string": [
   *     {
   *       "#text": "1.0.3",
   *     },
   *   ],
   * }
   * ```
   */
  private isStringNode = (node: PlistNode): node is {string: [{'#text': string | number}]} => {
    return !!node.string
  }
}

export const parsePlist = (plistPath: string): PlistContent => {
  const xmlFileContentString = readFileSync(plistPath).toString()
  const validationOutput = XMLValidator.validate(xmlFileContentString)

  if (validationOutput !== true) {
    throw new Error(validationOutput.err.msg)
  }

  const xmlParser = new XMLParser({preserveOrder: true})
  const [_xmlHeader, plistDeclaration] = xmlParser.parse(xmlFileContentString)

  const plistContent = plistDeclaration.plist[0].dict as PlistNode[]

  return new PlistContent(plistContent)
}
