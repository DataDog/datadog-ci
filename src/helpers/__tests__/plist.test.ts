import {parsePlist} from '../plist'

describe('plist util', () => {
  describe('parsePlist', () => {
    afterEach(() => {
      delete process.env.EXECUTABLE_NAME
    })
    it('parses the content of the info plist file', () => {
      const plist = parsePlist('src/helpers/__tests__/plist-fixtures/Info.plist')
      expect(plist.getContent()).toMatchSnapshot()
    })

    it('returns the value for a property', () => {
      const plist = parsePlist('src/helpers/__tests__/plist-fixtures/Info.plist')
      expect(plist.getPropertyValue('CFBundleShortVersionString')).toBe('1.0.4')
      expect(plist.getPropertyValue('CFBundleVersion')).toBe(12)
    })

    it('returns the value for an env variable', () => {
      process.env.EXECUTABLE_NAME = 'executable name'
      const plist = parsePlist('src/helpers/__tests__/plist-fixtures/Info.plist')
      expect(plist.getPropertyValue('CFBundleExecutable')).toBe('executable name')
    })

    it('returns an empty string for an env variable that is not declared', () => {
      const plist = parsePlist('src/helpers/__tests__/plist-fixtures/Info.plist')
      expect(() => plist.getPropertyValue('CFBundleExecutable')).toThrow(
        "Environment variable $(EXECUTABLE_NAME) for key CFBundleExecutable wasn't found."
      )
    })

    it('throws an error if a property does not exist', () => {
      const plist = parsePlist('src/helpers/__tests__/plist-fixtures/Info.plist')
      expect(() => plist.getPropertyValue('NonExistingValue')).toThrow('Property not found')
    })

    it('throws an error if a property is not a string', () => {
      const plist = parsePlist('src/helpers/__tests__/plist-fixtures/Info.plist')
      expect(() => plist.getPropertyValue('CFBundleURLTypes')).toThrow(
        'Property is not a string, this is not supported yet'
      )
    })

    it('throws an error if the file does not exist', () => {
      expect(() => parsePlist('non-existing-file')).toThrow(
        "ENOENT: no such file or directory, open 'non-existing-file'"
      )
    })
    it('throws an error if the file is not correctly formatted', () => {
      expect(() => parsePlist('src/helpers/__tests__/plist-fixtures/BadlyFormatted.plist')).toThrow(
        "Expected closing tag 'string' (opened in line 7, col 2) instead of closing tag 'dict'."
      )
    })
  })
})
