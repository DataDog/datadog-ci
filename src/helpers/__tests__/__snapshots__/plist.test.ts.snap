// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plist util parsePlist parses the content of the info plist file 1`] = `
Array [
  Object {
    "key": Array [
      Object {
        "#text": "CFBundleDevelopmentRegion",
      },
    ],
  },
  Object {
    "string": Array [
      Object {
        "#text": "$(DEVELOPMENT_LANGUAGE)",
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "CFBundleDisplayName",
      },
    ],
  },
  Object {
    "string": Array [
      Object {
        "#text": "sample-app",
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "CFBundleExecutable",
      },
    ],
  },
  Object {
    "string": Array [
      Object {
        "#text": "$(EXECUTABLE_NAME)",
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "CFBundleIdentifier",
      },
    ],
  },
  Object {
    "string": Array [
      Object {
        "#text": "$(PRODUCT_BUNDLE_IDENTIFIER)",
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "CFBundleInfoDictionaryVersion",
      },
    ],
  },
  Object {
    "string": Array [
      Object {
        "#text": 6,
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "CFBundleName",
      },
    ],
  },
  Object {
    "string": Array [
      Object {
        "#text": "$(PRODUCT_NAME)",
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "CFBundlePackageType",
      },
    ],
  },
  Object {
    "string": Array [
      Object {
        "#text": "$(PRODUCT_BUNDLE_PACKAGE_TYPE)",
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "CFBundleShortVersionString",
      },
    ],
  },
  Object {
    "string": Array [
      Object {
        "#text": "1.0.4",
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "CFBundleSignature",
      },
    ],
  },
  Object {
    "string": Array [
      Object {
        "#text": "????",
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "CFBundleURLTypes",
      },
    ],
  },
  Object {
    "array": Array [
      Object {
        "dict": Array [
          Object {
            "key": Array [
              Object {
                "#text": "CFBundleURLSchemes",
              },
            ],
          },
          Object {
            "array": Array [
              Object {
                "string": Array [
                  Object {
                    "#text": "myapp",
                  },
                ],
              },
              Object {
                "string": Array [
                  Object {
                    "#text": "com.sample.app",
                  },
                ],
              },
            ],
          },
        ],
      },
      Object {
        "dict": Array [
          Object {
            "key": Array [
              Object {
                "#text": "CFBundleURLSchemes",
              },
            ],
          },
          Object {
            "array": Array [
              Object {
                "string": Array [
                  Object {
                    "#text": "exp+sample-app",
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "CFBundleVersion",
      },
    ],
  },
  Object {
    "string": Array [
      Object {
        "#text": 12,
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "LSRequiresIPhoneOS",
      },
    ],
  },
  Object {
    "true": Array [],
  },
  Object {
    "key": Array [
      Object {
        "#text": "NSAppTransportSecurity",
      },
    ],
  },
  Object {
    "dict": Array [
      Object {
        "key": Array [
          Object {
            "#text": "NSAllowsArbitraryLoads",
          },
        ],
      },
      Object {
        "true": Array [],
      },
      Object {
        "key": Array [
          Object {
            "#text": "NSExceptionDomains",
          },
        ],
      },
      Object {
        "dict": Array [
          Object {
            "key": Array [
              Object {
                "#text": "localhost",
              },
            ],
          },
          Object {
            "dict": Array [
              Object {
                "key": Array [
                  Object {
                    "#text": "NSExceptionAllowsInsecureHTTPLoads",
                  },
                ],
              },
              Object {
                "true": Array [],
              },
            ],
          },
        ],
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "UILaunchStoryboardName",
      },
    ],
  },
  Object {
    "string": Array [
      Object {
        "#text": "SplashScreen",
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "UIRequiredDeviceCapabilities",
      },
    ],
  },
  Object {
    "array": Array [
      Object {
        "string": Array [
          Object {
            "#text": "armv7",
          },
        ],
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "UIRequiresFullScreen",
      },
    ],
  },
  Object {
    "false": Array [],
  },
  Object {
    "key": Array [
      Object {
        "#text": "UIStatusBarStyle",
      },
    ],
  },
  Object {
    "string": Array [
      Object {
        "#text": "UIStatusBarStyleDefault",
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "UISupportedInterfaceOrientations",
      },
    ],
  },
  Object {
    "array": Array [
      Object {
        "string": Array [
          Object {
            "#text": "UIInterfaceOrientationPortrait",
          },
        ],
      },
      Object {
        "string": Array [
          Object {
            "#text": "UIInterfaceOrientationPortraitUpsideDown",
          },
        ],
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "UISupportedInterfaceOrientations~ipad",
      },
    ],
  },
  Object {
    "array": Array [
      Object {
        "string": Array [
          Object {
            "#text": "UIInterfaceOrientationPortrait",
          },
        ],
      },
      Object {
        "string": Array [
          Object {
            "#text": "UIInterfaceOrientationPortraitUpsideDown",
          },
        ],
      },
      Object {
        "string": Array [
          Object {
            "#text": "UIInterfaceOrientationLandscapeLeft",
          },
        ],
      },
      Object {
        "string": Array [
          Object {
            "#text": "UIInterfaceOrientationLandscapeRight",
          },
        ],
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "UIUserInterfaceStyle",
      },
    ],
  },
  Object {
    "string": Array [
      Object {
        "#text": "Automatic",
      },
    ],
  },
  Object {
    "key": Array [
      Object {
        "#text": "UIViewControllerBasedStatusBarAppearance",
      },
    ],
  },
  Object {
    "false": Array [],
  },
]
`;
